README.md
# [Baremetrics](https://baremetrics.com/) Date Range Picker
_[Baremetrics](https://baremetrics.com) provides zero-setup subscription analytics & insights for Stripe, Braintree and Recurly. **[Get started today!](https://baremetrics.com)**_

---

The Baremetrics date range picker is a simplified solution for selecting both date ranges and single dates all from a single calender view. There aren't a billion options but the code is pretty basic and modular so feel free to edit however to meet your own needs.

Design by [Chris Meeks](https://dribbble.com/ChrisMeeks)  
Code by [Tyler van der Hoeven](https://github.com/tyvdh)

[View a demo](http://baremetrics.github.io/calendar/)  
[View in a live production app](https://demo.baremetrics.com/)

![](https://tyler.link/bqs5/Screen%20Shot%202015-07-02%20at%201.29.07%20PM.png)
![](https://tyler.link/br0F/Screen%20Shot%202015-07-02%20at%201.29.28%20PM.png)
![](https://tyler.link/bqV5/Screen%20Shot%202015-07-02%20at%201.32.15%20PM.png)

## Installing

Using the date picker is pretty simple, you've just got to make a couple choices and include a couple settings.
Create a div with a `daterange` class and then either the `daterange--double` or `daterange--single` classname for targeting the specific calendar you'd like to generate.

```html
<div class="daterange daterange--double"></div>
<div class="daterange daterange--single"></div>
```

Next you've just gotta create a `new Calendar` instance.

```js
new Calendar({
  element: $('.daterange--single'),
  current_date: '2015-06-15',
  format: {input: 'M/D/YYYY'},
  required: false
});

new Calendar({
  element: $('.daterange--double'),
  earliest_date: '2000-01-1',
  latest_date: moment(),
  start_date: '2015-05-01',
  end_date: '2015-05-31',
  callback: function() {
    var start = moment(this.start_date).format('ll'),
        end = moment(this.end_date).format('ll');

    console.debug('Start Date: '+ start +'\nEnd Date: '+ end);
  }
});
```

### Base Calendar Params
- **element** _\*required_ `[jQuery DOM object]`
  - jQuery DOM object of the calendar div you're working on
- **earliest_date** `[date YYYY-MM-DD]`
  - The earliest date to show in the calendar
- **latest_date** `[date YYYY-MM-DD]`
  - The latest date to show in the calendar
- **callback** `[function]`
  - A function for whenever a new date is saved
  - Inside you have access to object variables like `this.earliest_date` and `this.latest_date` for doing things with your calendar's dates
- **format** `[object]`
  - Object containing formatting strings for.. you guessed it.. formatting your dates
  ```js
    format: {
      input: 'MMMM D, YYYY', // Format for the input fields
      jump_month: 'MMMM', // Format for the month switcher
      jump_year: 'YYYY' // Format for the year switcher
    }
  ```
- **days_array** `[array]`
  - Array of the 7 strings you'd like to represent your days in the calendar
  ```js
    days_array: ['S','M','T','W','T','F','S']
  ```

### Single Calendar Params
- **current_date** `[date YYYY-MM-DD]`
  - The date to start the calendar on
- **required** `[boolean]`
  - Toggle if this field must have always have a valid selected date
- **placeholder** `[string]`
  - Set placeholder text (note this will only apply if the required key is set to `false`). The default will be whatever moment date format you're using. (i.e. 'M/D/YYYY')

### Double Calendar Params
- **start_date** `[date YYYY-MM-DD]`
  - The date to start the selection on for the calendar
- **end_date** `[date YYYY-MM-DD]`
  - The date to end the selection on for the calendar
- **same_day_range** `[boolean]`
  - Allow a range selection of a single day
- **format** `[preset key in format object] // see above`
  - The double calendar adds the `preset` key to the format object for formatting the preset dates in the preset dropdown
- **presets** `[boolean] or [object]`
  - If you don't want to show the preset link just set this to `false` otherwise the default is true which will just give you a basic preset of.. yep.. presets. BOOM!
  - Otherwise, if you want to customize it up you can include an array of preset objects. Something like:
  ```js
    presets: [{
      label: 'Last month',
      start: moment().subtract(1, 'month').startOf('month'),
      end: moment().subtract(1, 'month').endOf('month')
    },{
      label: 'Last year',
      start: moment().subtract(1, 'year').startOf('year'),
      end: moment().subtract(1, 'year').endOf('year')
    }]
  ```

---

## Developing

I've included my signature gulpfile too so be sure and take a look at that as well.

```bash
$ cd <project directory>
$ npm install
$ gulp
```

I also use [pow](http://pow.cx/) and the [powder gem](https://github.com/Rodreegez/powder) to run my local dev environments but however you plan on wrangling that the gulpfile turns on a livereload server so as long as you have the files serving somehow any changes you make will show up instantly.

## Dependencies
- [jQuery](https://jquery.com/)
- [MomentJS](http://momentjs.com/)


changelog.md
## 1.0.14 Update
- Fixed calendar close clicking issues. It's harder than you might think

## 1.0.13 Update
- Microsoft Edge compatibility updates

## 1.0.12 Update
- Mostly just a few key PR inclusions
- Few small tweaks and minor bug fixes

## 1.0.11 Update
- Third party library updates
- Typos and code formatting. Because I'm paranoid

## 1.0.10 Update
- Better iOS support. The date inputs are no longer editable. Having a keyboard popup for editing dates was a lot slower then just tapping stuff. So we cut it. Thanks [azaytsev-csr](https://github.com/azaytsev-csr)!
- Minor tweaks and outlier bug fixes Thanks [Paul Ryan](https://github.com/paullryan)!
- More consistent moment usage. Cleaned up the code base quite a bit. Thanks [Theodore Brown](https://github.com/theodorejb)!
- Fire callback on single calendar when value is deleted [](https://github.com/Baremetrics/calendar/issues/68)

## 1.0.9 Update
- Don't automatically include the "This month" preset in with custom presets (sorry guys)
- If required is set to `false` for a date range but you've included a `current_date` value the input will pre-fill with that date rather then defaulting to the placeholder

## 1.0.8 Update
- Pulled in a couple pull requests to fix some locale issues
- Couple small bug fixes and performance improvements

## 1.0.7 Update
Originally the compiled CSS file had both my demo app CSS and the calendar CSS all mashed together. Apparently some folks found that to be annoying. (Hint: It was me) This is no longer the case.

## 1.0.6 Update
Custom presets!!

- **presets** `[boolean] or [object]`
  - If you don't want to show the preset link just set this to `false` otherwise the default is true which will just give you a basic preset of.. yep.. presets. BOOM!
  - Otherwise, if you want to customize it up you can include an array of preset objects. Something like:
  ```js
    presets: [{
      label: 'Last month',
      start: moment().subtract(1, 'month').startOf('month'),
      end: moment().subtract(1, 'month').endOf('month')
    },{
      label: 'Last year',
      start: moment().subtract(1, 'year').startOf('year'),
      end: moment().subtract(1, 'year').endOf('year')
    }]
  ```

We also added custom formatting for the placeholder text on the single date picker if the date is not required.

- **placeholder** `[string]`
  - Set placeholder text (note this will only apply if the required key is set to `false`)

Last but not least we also added support for locale based first day of the week.

## 1.0.5 Update
Version 1.0.5 sees two major improvements.

The first is you can now add a `required` key for the single date picker if you'd like to break the default and generate a picker with a placeholder rather then a predated input.

```js
new Calendar({
  element: $('.daterange--single'),
  current_date: 'June 15, 2015',
  format: {input: 'M/D/YYYY'},
  required: false
});
```

![](http://tyler.link/dw0p/Screen%20Shot%202015-11-27%20at%202.36.06%20PM.png)

Note the input format will be the placeholder. Cool huh?

The second is that *finally* the picker's callback function will only fire if the dates have actually changed. Before the function would fire regardless of what dates were chosen. So dumb. We're smarter now.

## 1.0.4 Update
- Formats! You can now format all the output dates via the format object in the `new Calendar` call (Thanks [Derk-Jan Karrenbeld](https://github.com/SleeplessByte))
```js
new Calendar({
  ...
  format = {
    input: 'MMMM D, YYYY', // Format for the input fields
    preset: 'll', // Format for the preset dates
    jump_month: 'MMMM', // Format for the month switcher
    jump_year: 'YYYY' // Format for the year switcher
  }
  ...
  }
});
```
- Days array. You can now include your own days array if you'd like to change the values at the top of the calendar
```js
new Calendar({
  ...
  days_array: ['Su','Mo','Tu','We','Th','Fr','Sa']
  ...
  }
});
```
- Added support for manual date entry outside the range.. will now clamp back/forward to latest/earliest dates rather then just resetting. Yay!
- On the double calendar, callbacks now submit on a click event outside the calendar if a valid date has been chosen for the first date. No need to pick the same second date just to submit the changes!
- A few more minor bug fixes and improvements

## 1.0.3 Update

- Added basic shorthand keyword support
  - `now` or `today` to set date to today's date
  - `earliest` or `latest` to set date to the preset earliest or latest dates respectively
  - `[number] [interval] [ago/ahead]` to advance or regress the date based off the currently selected date (i.e. `1 month ago` on a date of August 1 would result in a date of July 1)
- You may now use plain date strings in the Calendar object
- Fixed a bug where auto selecting a week range after or before the current selection was getting suppressed and reset to the previous dates
- Added support for selecting a range of single day with the `same_day_range` boolean in the double calendar call
- Cleaned and dried out a couple verbose functions

## 1.0.2 Update

- Added license and repo info to the package.json
- Removed underscore dependency
- Added the github page publisher function to the gulpfile
- Added a switch statement for keycode checking (Thanks again [Theodore Brown](https://github.com/theodorejb))
- Added the `'use strict'` label to the Calendar.js file
- Cleaned out some unused gulp functions
- Compressed the Sass folder to a single level

## 1.0.1 Update

- Removed global CSS reset file
- Fixed issue with presets not respecting earliest date
- Registered the package with Bower `bower install BaremetricsCalendar` (Thanks [Agustin Diaz](https://github.com/HiroAgustin))
- Shifted code base to support UMD support (Thanks [Derrick Reimer](https://github.com/djreimer))
- Added up/down keyboard support
  - day = keystroke up/down
  - week = hold shift + keystroke up/down
  - month = hold meta + keystroke up/down
- Fixed a couple minor typos hither and thither


gulpfile.js
var gulp = require('gulp'),
    ghPages = require('gulp-gh-pages'),
    sass = require('gulp-sass'),
    autoprefixer = require('gulp-autoprefixer'),
    cleancss = require('gulp-clean-css'),
    uglify = require('gulp-uglify'),
    livereload = require('gulp-livereload'),
    newer = require('gulp-newer');


// ROOT TASKS // ---------------------------------------------------------
// Main style task  
function css(cb) {
  gulp.src('dev/sass/**/**.scss')
    .pipe(sass())
    .on('error', handleError)
    .pipe(autoprefixer({cascade: false,  browsers: ['last 10 versions']})) // auto prefix
    .pipe(cleancss()) // minify everything
    .pipe(gulp.dest('public/css'));

  cb();
}

// Main Javascript task
function js(cb) {
  gulp.src('dev/js/**/*.js')
    .pipe(newer('public/js'))
    // .pipe(uglify())
    .on('error', handleError)
    .pipe(gulp.dest('public/js'));

  cb();
}

// Publish github page
function deploy(cb) {
  gulp.src('./public/**/*')
    .pipe(ghPages());

  cb();
}


// FUNCTIONS // ---------------------------------------------------------
// Watch function
function watch(cb) {
  gulp.watch('dev/js/**/*.js', js);
  gulp.watch('dev/sass/**/*.scss', css);
  // gulp.watch('dev/img/**/*.{jpg,jpeg,png,gif,svg,ico}', img);
 
  livereload.listen();
  gulp.watch(['public/*.html', 'public/js/**/*.js', 'public/img/**/*.{jpg,jpeg,png,gif,svg,ico}', 'public/css/*.css']).on('change', livereload.changed);

  cb();
}

// Error reporting function
function handleError(err) {
  console.log(err.toString());
  this.emit('end');
}

// Default function
exports.default = gulp.series(js, css, watch)

exports.css = css
exports.js = js
exports.deploy = deploy

dev/js/app.js
var ss = new Calendar({
  element: $('.daterange--single'),
  format: {input: 'M/D/YYYY'},
  placeholder: 'Select a Date',
  required: false,
  callback: function() {
    console.log('Single calendar callback triggered');
  }
});

var dd = new Calendar({
  element: $('.one'),
  earliest_date: '2000-01-01',
  latest_date: moment(),
  start_date: moment().subtract(29, 'days'),
  end_date: moment(),
  callback: function() {
    var start = moment(this.start_date).format('ll'),
        end = moment(this.end_date).format('ll'),
        preset = this.preset_label;

    console.debug('Start Date: ' + start + '\nEnd Date: ' + end + '\nPreset: ' + preset);
  }
});

new Calendar({
  element: $('.two'),
  earliest_date: '2000-01-01',
  latest_date: moment(),
  start_date: moment().subtract(29, 'days'),
  end_date: moment(),
  presets: false,
  callback: function() {
    var start = moment(this.start_date).format('ll'),
        end = moment(this.end_date).format('ll'),
        preset = this.preset_label;

    console.debug('Start Date: ' + start + '\nEnd Date: ' + end + '\nPreset: ' + preset);
  }
});

new Calendar({
  element: $('.three'),
  earliest_date: '2000-01-01',
  latest_date: moment(),
  start_date: moment().subtract(29, 'days'),
  end_date: moment(),
  presets: [{
    label: 'Last 30 days',
    start: moment().subtract(29, 'days'),
    end: moment()
  },{
    label: 'Last month',
    start: moment().subtract(1, 'month').startOf('month'),
    end: moment().subtract(1, 'month').endOf('month')
  },{
    label: 'Last year',
    start: moment().subtract(1, 'year').startOf('year'),
    end: moment().subtract(1, 'year').endOf('year')
  }],
  callback: function() {
    var start = moment(this.start_date).format('ll'),
        end = moment(this.end_date).format('ll'),
        preset = this.preset_label;

    console.debug('Start Date: ' + start + '\nEnd Date: ' + end + '\nPreset: ' + preset);
  }
});


dev/js/base.js
function Calendar(settings) {
  this.calIsOpen = false;
  this.presetIsOpen = false;
  this.element = null;
  this.earliest = settings ? settings.earliest : new Date('January 1, 1900');
  this.latest = settings ? settings.latest : new Date('December 31, 2900');
  this.start_date = settings ? settings.start : null;
  this.end_date = settings ? settings.end : null;
  this.current_date = null;

  var self = this;

  $('.dr-presets').click(function() {
    self.presetToggle();
  });

  $('.dr-list-item').click(function() {
    var range = $('.dr-item-aside', this).html().split('–');

    self.start_date = new Date(range[0]);
    self.end_date = new Date(range[1]);
    self.calendarSetDates();
    self.presetToggle();
    self.calendarSaveDates();
  });

  $('.dr-date').on({
    'click': function() {
      self.calendarOpen(this);
    },

    'keyup': function(event) {
      if (event.keyCode == 9 && !self.calIsOpen && !self.start_date && !self.end_date)
        self.calendarOpen(this);
    },

    'keydown': function(event) {
      if (event.keyCode == 13) { // Enter
        event.preventDefault();
        self.calendarCheckDates();
        self.calendarSetDates();
        self.calendarClose('force');
        
        if ($(self.element).hasClass('dr-date-start') || 
            $(self.element).hasClass('dr-date-end'))
          self.calendarSaveDates();
      }

      if (event.keyCode == 27) { // ESC
        self.calendarSetDates();
        self.calendarClose('force');
      }

      if (event.keyCode == 9) { // Tab
        if ($(self.element).hasClass('dr-date-start')) {
          event.preventDefault();
          $('.dr-date-end').trigger('click');
        } else if ($(self.element).hasClass('dr-date-end')) {
          event.preventDefault();
          self.calendarCheckDates();
          self.calendarSaveDates();
          self.calendarClose('force');
        } else {
          self.calendarCheckDates();
          self.calendarSetDates();
          self.calendarClose('force');
        }
      }
    }
  });

  $('.dr-month-switcher i').click(function() {
    var m = $('.dr-month-switcher span').html();
    var y = $('.dr-year-switcher span').html();
    var back = moment(new Date(m +' 1, '+ y)).subtract(1, 'month');
    var forward = moment(new Date(m +' 1, '+ y)).add(1, 'month').startOf('day');

    if ($(this).hasClass('icon-left')) {  
      $(this).parent().find('span').html(back.format('MMMM'));
      self.calendarOpen(self.element, back);
    } else if ($(this).hasClass('icon-right')) {
      $(this).parent().find('span').html(forward.format('MMMM'));
      self.calendarOpen(self.element, forward);
    }
  });

  $('.dr-year-switcher i').click(function() {
    var m = $('.dr-month-switcher span').html();
    var y = $('.dr-year-switcher span').html();
    var back = moment(new Date(m +' 1, '+ y)).subtract(1, 'year');
    var forward = moment(new Date(m +' 1, '+ y)).add(1, 'year').startOf('day');

    if ($(this).hasClass('icon-left')) {  
      $(this).parent().find('span').html(back.format('YYYY'));
      self.calendarOpen(self.element, back);
    } else if ($(this).hasClass('icon-right')) {
      $(this).parent().find('span').html(forward.format('YYYY'));
      self.calendarOpen(self.element, forward);
    }
  });

  $('.dr-dates-dash').click(function() {
    $('.dr-date-start').trigger('click');
  });

  $('.daterange').click(function(event) {
    $('html').one('click',function() {
      if (self.presetIsOpen)
        self.presetToggle();

      if (self.calIsOpen) {
        self.calendarSetDates();
        self.calendarClose('force');
      }
    });

    event.stopPropagation();
  });

  $('.daterange, .dr-date').focus(function(event) {
    $('html').one('click',function() {
      if (self.calIsOpen) {
        self.calendarSetDates();
        self.calendarClose('force');
      }
    });

    event.stopPropagation();
  });
}


Calendar.prototype.presetToggle = function() {
  if (this.presetIsOpen == false) {
    this.presetIsOpen = true;
    this.presetCreate();
  } else if (this.presetIsOpen) {
    this.presetIsOpen = false;
  }

  if (this.calIsOpen == true)
    this.calendarClose();

  $('.dr-preset-list').slideToggle(200);
  $('.dr-input').toggleClass('active');
  $('.dr-presets').toggleClass('active');
}


Calendar.prototype.presetCreate = function() {
  var self = this;
  var date = this.latest;

  var s = new Date($('.dr-date-start').html());
  var e = new Date($('.dr-date-end').html());
  this.start_date = s == 'Invalid Date' ? this.start_date : s;
  this.end_date = e == 'Invalid Date' ? this.end_date : e;

  $('.dr-list-item').each(function() {
    var month_count = $(this).data('months');
    var last_day = moment(date).endOf('month').startOf('day');
    var is_last_day = last_day.isSame(date);
    var first_day;

    if (!is_last_day)
      last_day = moment(date).subtract(1, 'month').endOf('month').startOf('day');

    if (typeof month_count == 'number') {
      first_day = moment(date).subtract(is_last_day ? month_count - 1 : month_count, 'month').startOf('month');

      if (month_count == 12)
        first_day = moment(date).subtract(is_last_day ? 12 : 13, 'month').endOf('month').startOf('day');
    } else if (month_count == 'all') {
      first_day = moment(self.earliest);
      last_day = moment(self.latest);
    } else {
      first_day = moment(self.latest).subtract(30, 'day');
      last_day = moment(self.latest);
    }

    $('.dr-item-aside', this).html(first_day.format('ll') +' – '+ last_day.format('ll'));
  });
}


Calendar.prototype.calendarSetDates = function() {
  $('.dr-date-start').html(moment(this.start_date).format('MMMM D, YYYY'));
  $('.dr-date-end').html(moment(this.end_date).format('MMMM D, YYYY'));

  if (!this.start_date && !this.end_date) {
    var old_date = $('.dr-date').html();
    var new_date = moment(this.current_date).format('MMMM D, YYYY');

    if (old_date != new_date)
      $('.dr-date').html(new_date);
  }
}


Calendar.prototype.calendarSaveDates = function() {
  var start = moment(this.start_date).format("YYYY-MM-DD"),
      end = moment(this.end_date).format("YYYY-MM-DD"),
      url = "?start_date="+ start +"&end_date="+ end,
      full = (window.location.pathname + url + window.location.hash);
  
  window.setTimeout(function() {
    window.location = full;
  }, 200);
}


Calendar.prototype.calendarCheckDates = function() {
  var regex = /(?!<=\d)(st|nd|rd|th)/;
  var s = $('.dr-date-start').html();
  var e = $('.dr-date-end').html();
  var c = $(this.element).html();
  var s_array = [];
  var e_array = [];
  var c_array = [];

  if (s) {
    s = s.replace(regex, '');
    s_array = s.split(' ');
  }

  if (e) {
    e = e.replace(regex, '');
    e_array = e.split(' ');
  }

  if (c) {
    c = c.replace(regex, '');
    c_array = c.split(' ');
  }

  if (s_array.length == 2) {
    s_array.push(moment().format('YYYY'))
    s = s_array.join(' ');
  }

  if (e_array.length == 2) {
    e_array.push(moment().format('YYYY'))
    e = e_array.join(' ');
  }

  if (c_array.length == 2) {
    c_array.push(moment().format('YYYY'))
    c = c_array.join(' ');
  }

  s = new Date(s);
  e = new Date(e);
  c = new Date(c);

  if (moment(s).isAfter(e) || 
      moment(e).isBefore(s) || 
      moment(s).isSame(e) ||
      moment(s).isBefore(this.earliest) ||
      moment(e).isAfter(this.latest)) {
    return this.calendarSetDates();
  }

  this.start_date = s == 'Invalid Date' ? this.start_date : s;
  this.end_date = e == 'Invalid Date' ? this.end_date : e;
  this.current_date = c == 'Invalid Date' ? this.current_date : c;
}


Calendar.prototype.calendarOpen = function(element, switcher) {
  var self = this;
  var other;
  var cal_width = $('.dr-dates').innerWidth() - 8;

  this.element = element || this.element;

  if (this.presetIsOpen == true)
    this.presetToggle();

  if (this.calIsOpen == true)
    this.calendarClose(switcher ? 'switcher' : undefined);

  this.calendarCheckDates();
  this.calendarCreate(switcher);
  this.calendarSetDates();

  var next_month = moment(switcher || this.current_date).add(1, 'month').startOf('month').startOf('day');
  var past_month = moment(switcher || this.current_date).subtract(1, 'month').endOf('month');
  var next_year = moment(switcher || this.current_date).add(1, 'year').startOf('month').startOf('day');
  var past_year = moment(switcher || this.current_date).subtract(1, 'year').endOf('month');

  $('.dr-month-switcher span').html(moment(switcher || this.current_date).format('MMMM'));
  $('.dr-year-switcher span').html(moment(switcher || this.current_date).format('YYYY'));

  $('.dr-switcher i').removeClass('dr-disabled');

  if (next_month.isAfter(this.latest))
    $('.dr-month-switcher .icon-right').addClass('dr-disabled');

  if (past_month.isBefore(this.earliest))
    $('.dr-month-switcher .icon-left').addClass('dr-disabled');

  if (next_year.isAfter(this.latest))
    $('.dr-year-switcher .icon-right').addClass('dr-disabled');

  if (past_year.isBefore(this.earliest))
    $('.dr-year-switcher .icon-left').addClass('dr-disabled');

  $('.dr-day').on({
    mouseenter: function() {
      var element = $(this);
      var start_date = moment(self.start_date);
      var end_date = moment(self.end_date);
      var current_date = moment(self.current_date);

      if (start_date.isSame(current_date)) {
        element.addClass('dr-hover dr-hover-before');
        $('.dr-start').css({'border': 'none', 'padding-left': '0.3125rem'});
        setMaybeRange('start');
      }

      if (end_date.isSame(current_date)) {
        element.addClass('dr-hover dr-hover-after');
        $('.dr-end').css({'border': 'none', 'padding-right': '0.3125rem'});
        setMaybeRange('end');
      }

      if (!self.start_date && !self.end_date)
        element.addClass('dr-maybe');

      $('.dr-selected').css('background-color', 'transparent');

      function setMaybeRange(type) {
        other = undefined;

        $.each(_.range(6 * 7), function(i) {
          var next = element.next().data('date');
          var prev = element.prev().data('date');
          var curr = element.data('date');

          if (!curr)
            return false;

          if (!prev)
            prev = curr;

          if (!next)
            next = curr;

          if (type == 'start')
            if (moment(next).isSame(self.end_date))
              return false;

          if (type == 'end')
            if (moment(prev).isSame(self.start_date))
              return false;


          if (type == 'start') {
            if (moment(curr).isAfter(self.end_date)) {
              other = other || moment(curr).add(6, 'day').startOf('day');

              if (i > 5 || (next ? moment(next).isAfter(self.latest) : false)) {
                $(element).addClass('dr-end');
                other = moment(curr);
                return false;
              }
            }
          }

          if (type == 'end') {
            if (moment(curr).isBefore(self.start_date)) {
              other = other || moment(curr).subtract(6, 'day');

              if (i > 5 || (prev ? moment(prev).isBefore(self.earliest) : false)) {
                $(element).addClass('dr-start');
                other = moment(curr);
                return false;
              }
            }
          } 


          if (type == 'start')
            element = element.next().addClass('dr-maybe');

          if (type == 'end')
            element = element.prev().addClass('dr-maybe');
        });
      }
    },
    mouseleave: function() {
      if ($(this).hasClass('dr-hover-before dr-end'))
        $(this).removeClass('dr-end');

      if ($(this).hasClass('dr-hover-after dr-start'))
        $(this).removeClass('dr-start');

      $(this).removeClass('dr-hover dr-hover-before dr-hover-after');
      $('.dr-start, .dr-end').css({'border': '', 'padding': ''});
      $('.dr-maybe:not(.dr-current)').removeClass('dr-start dr-end');
      $('.dr-day').removeClass('dr-maybe');
      $('.dr-selected').css('background-color', '');
    },
    mousedown: function() {
      var date = $(this).data('date');
      var string = moment(date).format('MMMM D, YYYY');

      if (other) {
        $('.dr-date')
          .not(self.element)
          .html(other.format('MMMM D, YYYY'));
      }

      $(self.element).html(string);
      self.calendarOpen(self.element);

      if ($(self.element).hasClass('dr-date-start')) {
        $('.dr-date-end').trigger('click');
      } else if ($(self.element).hasClass('dr-date-end')) {
        self.calendarSaveDates();
        self.calendarClose('force');
      } else {
        self.calendarClose('force');
      }
    }
  });

  $('.dr-calendar')
    .css('width', cal_width)
    .slideDown(200);
  $('.dr-input').addClass('active');
  $(element).addClass('active').focus();

  this.calIsOpen = true;
}


Calendar.prototype.calendarClose = function(type) {
  if (!this.calIsOpen || this.presetIsOpen || type == 'force') {
    $('.dr-date-start, .dr-date-end').blur();
    $('.dr-calendar').slideUp(200, function() {
      $('.dr-day').remove();
    });
  } else {
    $('.dr-day').remove();
  }

  if (type == 'switcher') {
    return false;
  }

  $('.dr-input, .dr-date').removeClass('active');

  this.calIsOpen = false;
}


Calendar.prototype.calendarArray = function(start, end, current, switcher) {
  var self = this;
  var current = current || start || end;

  var first_day = moment(switcher || current).startOf('month');
  var last_day = moment(switcher || current).endOf('month');

  var current_month = {
    start: {
      day: +first_day.format('d'), 
      str: +first_day.format('D')
    }, 
    end: {
      day: +last_day.format('d'), 
      str: +last_day.format('D')
    }
  }


  // Beginning faded dates
  var d = undefined;

  var start_hidden = _.map(_.range(current_month.start.day), function() {
    if (d == undefined) {
      d = moment(first_day);
    } d = d.subtract(1, 'day');

    return {
      str: +d.format('D'),
      start: d.isSame(start),
      end: d.isSame(end),
      current: d.isSame(current),
      selected: d.isBetween(start, end),
      date: d.toISOString(),
      outside: d.isBefore(self.earliest),
      fade: true
    }
  }).reverse();


  // Leftover faded dates
  var leftover = (6 * 7) - (current_month.end.str + start_hidden.length);
  d = undefined;

  var end_hidden = _.map(_.range(leftover), function() {
    if (d == undefined) {
      d = moment(last_day);
    } d = d.add(1, 'day').startOf('day');

    return {
      str: +d.format('D'),
      start: d.isSame(start),
      end: d.isSame(end),
      current: d.isSame(current),
      selected: d.isBetween(start, end),
      date: d.toISOString(),
      outside: d.isAfter(self.latest),
      fade: true
    }
  });


  // Actual visible dates
  d = undefined;

  var visible = _.map(_.range(current_month.end.str), function() {
    if (d == undefined) {
      d = moment(first_day);
    } else {
      d = d.add(1, 'day').startOf('day');
    }

    return {
      str: +d.format('D'),
      start: d.isSame(start),
      end: d.isSame(end),
      current: d.isSame(current),
      selected: d.isBetween(start, end),
      date: d.toISOString(),
      outside: d.isBefore(self.earliest) || d.isAfter(self.latest)
    }
  });


  return start_hidden.concat(visible, end_hidden);
}


Calendar.prototype.calendarCreate = function(switcher) {
  var array = this.calendarArray(this.start_date, this.end_date, this.current_date, switcher);

  _.each(array, function(d, i) {
    var classString = "dr-day";

    if (d.fade)
      classString += " dr-fade";

    if (d.start)
      classString += " dr-start";

    if (d.end)
      classString += " dr-end";

    if (d.current)
      classString += " dr-current";
    
    if (d.selected)
      classString += " dr-selected";

    if (d.outside)
      classString += " dr-outside";

    $('.dr-day-list').append('<li class="'+ classString +'" data-date="'+ d.date +'">'+ d.str +'</li>');
  });
}

if ($('.daterange--single').length) {
  new Calendar();
} else if ($('.daterange').length) {
  new Calendar({
    earliest: new Date($('.daterange').data('earliest')),
    latest: new Date($('.daterange').data('latest')),
    start: new Date($('.dr-date-start').html()),
    end: new Date($('.dr-date-end').html())
  });
}